src/main/java/io/numaproj/numaflow/function/FunctionServer.java

```diff
@@ -151,7 +155,23 @@ public <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(
                                 headers.get(FunctionConstants.DATUM_METADATA_WIN_START),
                                 FunctionConstants.WINDOW_END_TIME,
                                 headers.get(FunctionConstants.DATUM_METADATA_WIN_END));
-                return Contexts.interceptCall(context, call, headers, next);
+                ServerCall.Listener<ReqT> listener = Contexts.interceptCall(context, call, headers, next);
+                return new ForwardingServerCallListener.SimpleForwardingServerCallListener<>(listener) {
+                    @Override
+                    public void onHalfClose() {
+                        try {
+                            super.onHalfClose();
+                        } catch (RuntimeException ex) {
+                            handleException(ex, call, headers);
+                            throw ex;
+                        }
+                    }
+                    private void handleException(RuntimeException exception, ServerCall<ReqT, RespT> serverCall, Metadata headers) {
+                        // Currently, we only have application level exceptions.
+                        // Translate it to UNKNOWN status.
+                        serverCall.close(Status.UNKNOWN, headers);
+                    }
+                };
```

Comment:
Why do we need this? Can't this be achieved using the onError() method of the observer?

Reply:
Using the onError() works but it requires specific error handling in each case there is exception. This is to use a GRPC interceptor to build a central place for error handling, so any new exception can be added and handled here without adding extra logic in other places.

Reply:
I think your questions is orthogonal to whether using onError directly or interceptor. More for how to handle the errors then. For map, the error handling case isn't for application error inside the UDF? And for reduce, we need to send back error for the client to say if retry is needed to match go/python sdk behavior.